components/ws-daemon/pkg/controller/workspace_controller_test.go

```diff
@@ -0,0 +1,268 @@
+)
+
+var _ = Describe("WorkspaceController", func() {
+	Context("with regular workspace", func() {
+		It("should handle regular content init", func() {
+			name := uuid.NewString()
+
+			mockCtrl := gomock.NewController(GinkgoT())
+			defer mockCtrl.Finish()
+			ops := NewMockWorkspaceOperations(mockCtrl)
+
+			ops.EXPECT().InitWorkspace(gomock.Any(), gomock.Any()).Return("", nil).Times(1)
+			workspaceCtrl.operations = ops
+
+			_ = createSecret(fmt.Sprintf("%s-tokens", name), secretsNamespace)
+			ws := newWorkspace(name, workspaceNamespace, workspacev1.WorkspacePhaseCreating)
+			createWorkspace(ws)
+			updateObjWithRetries(k8sClient, ws, true, func(ws *workspacev1.Workspace) {
+				ws.Status.Phase = workspacev1.WorkspacePhaseCreating
+				ws.Status.Conditions = []metav1.Condition{}
+				ws.Status.Runtime = &workspacev1.WorkspaceRuntimeStatus{
+					NodeName: NodeName,
+				}
+			})
+
+			expectConditionEventually(ws, string(workspacev1.WorkspaceConditionContentReady), metav1.ConditionTrue, "InitializationSuccess")
+		})
+
+		It("should handle regular content backup", func() {
+			name := uuid.NewString()
+
+			mockCtrl := gomock.NewController(GinkgoT())
+			defer mockCtrl.Finish()
+			ops := NewMockWorkspaceOperations(mockCtrl)
+
+			gitStatus := &csapi.GitStatus{
+				Branch:               "main",
+				LatestCommit:         "991300e0cf199116685f25561702a145d40ae462",
+				UncommitedFiles:      []string{"git", "pod"},
+				TotalUncommitedFiles: 2,
+				UntrackedFiles:       []string{"kumquat"},
+				TotalUntrackedFiles:  1,
+				UnpushedCommits:      []string{"df591ed557c9afa8b6bcd1f51809d83d3f48fc43"},
+				TotalUnpushedCommits: 1,
+			}
+
+			ops.EXPECT().BackupWorkspace(gomock.Any(), gomock.Any()).Return(gitStatus, nil).Times(1)
+			ops.EXPECT().DeleteWorkspace(gomock.Any(), gomock.Any())
+			workspaceCtrl.operations = ops
+
+			_ = createSecret(fmt.Sprintf("%s-tokens", name), secretsNamespace)
+			ws := newWorkspace(name, workspaceNamespace, workspacev1.WorkspacePhaseCreating)
+			createWorkspace(ws)
+			markContentReady(ws)
+
+			expectConditionEventually(ws, string(workspacev1.WorkspaceConditionBackupComplete), metav1.ConditionTrue, "BackupComplete")
+			expectGitStatusEventually(ws, gitStatus)
+		})
+
+		It("should report backup failure", func() {
+			name := uuid.NewString()
+
+			mockCtrl := gomock.NewController(GinkgoT())
+			defer mockCtrl.Finish()
+			ops := NewMockWorkspaceOperations(mockCtrl)
+
+			ops.EXPECT().BackupWorkspace(gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("BOOM!")).Times(1)
+			ops.EXPECT().DeleteWorkspace(gomock.Any(), gomock.Any())
+			workspaceCtrl.operations = ops
+
+			_ = createSecret(fmt.Sprintf("%s-tokens", name), secretsNamespace)
+			ws := newWorkspace(name, workspaceNamespace, workspacev1.WorkspacePhaseCreating)
+			createWorkspace(ws)
+			markContentReady(ws)
+
+			expectConditionEventually(ws, string(workspacev1.WorkspaceConditionBackupFailure), metav1.ConditionTrue, "BackupFailed")
```

Comment:
maybe nice to check here that the condition message contains the failure reason as well